src/serde/room_state.rs

```
@@ -0,0 +1,136 @@
+//! A module to deserialize a RoomState struct from incorrectly specified v1
+//! send_join endpoint.
+//!
+//! For more information, see this [GitHub issue](https://github.com/matrix-org/matrix-doc/issues/2541).
+
+use std::fmt;
+
+use serde::{
+    de::{self, Deserializer, Error, SeqAccess, Visitor},
+    ser::{SerializeSeq, Serializer},
+};
+
+use crate::v1::join::create_join_event::RoomState;
+
+pub fn serialize<S>(room_state: &RoomState, serializer: S) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+{
+    let mut seq = serializer.serialize_seq(Some(2))?;
+    seq.serialize_element(&200)?;
+    seq.serialize_element(room_state)?;
+    seq.end()
+}
+
+pub fn deserialize<'de, D>(deserializer: D) -> Result<RoomState, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    deserializer.deserialize_seq(RoomStateVisitor)
+}
+
+struct RoomStateVisitor;
+
+impl<'de> Visitor<'de> for RoomStateVisitor {
+    type Value = RoomState;
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_str("Room State response wrapped in an array.")
+    }
+
+    fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+    where
+        A: SeqAccess<'de>,
+    {
+        let bogus = seq
+            .next_element::<u32>()?
+            .ok_or_else(|| de::Error::custom("Expected integer as first element"))?;
+        if bogus != 200 {
+            return Err(A::Error::invalid_value(
+                de::Unexpected::Unsigned(bogus.into()),
+                &self,
+            ));
+        }
```

I've checked the spec again and it never says that this integer has to be 200. We shouldn't fail deserialization if it's a different value. Let's just verify it is an integer, and then discard it.

---

OK. When I talked to Travis about it, he said that we should just set it to 200 for serialization, but it makes sense not to fail on deserialization if we don't care about the value.

---

Turns out you have to explicitly ignore any extra values as well.

---

Oh, I wonder whether we should do that then.. It feels weird. What do you think?

---

I feel like it's fine; accepting the minimum amount of correctness on input but requiring complete correctness on output seems like a good policy to have.

---

On the other hand, bad actors could send a giant payload, and we'd be executing extra statements for no reason. It's not unreasonable to expect homeservers to follow spec, so failing early is probably the better choice.